src/charm.py

```diff
@@ -15,30 +15,26 @@
 import ops
 import yaml
 from charms.grafana_agent.v0.cos_agent import COSAgentProvider
+from charms.operator_libs_linux.v2.snap import SnapError
 from ops.model import (
     ActiveStatus,
     BlockedStatus,
     ModelError,
     WaitingStatus,
 )
 
-from service import get_installed_snap_service, snap_install
+from service import SNAP_NAME, get_installed_snap_service, snap_install_or_refresh
 
 logger = logging.getLogger(__name__)
 
-# charm global constants
 RESOURCE_NAME = "openstack-exporter"
-
-# snap global constants
-SNAP_NAME = "golang-openstack-exporter"
-
 # Snap config options global constants
 # This is to match between openstack-exporter and the entry in clouds.yaml
 CLOUD_NAME = "openstack"
 # store the clouds.yaml where it's easily accessible by the openstack-exporter snap
 # This is the SNAP_COMMON directory for the exporter snap, which is accessible,
 # unversioned, and retained across updates of the snap.
-OS_CLIENT_CONFIG = "/var/snap/golang-openstack-exporter/common/clouds.yaml"
+OS_CLIENT_CONFIG = f"/var/snap/{SNAP_NAME}/common/clouds.yaml"
```

Comment:
What happens in an upgrade scenario? i.e. have you tested possible upgrade scenarios?

Let's say:

- the charm was deployed with a local resource of golang-openstack-exporter snap initially
- the charm gets upgraded to the charm revision with this code
- the resource was attached initially with the golang-openstack-exporter snap and it's still there
- new charm revision skips the installation of the new charmed-openstack-exporter snap from the store (I might be wrong here)

Can the old golang-openstack-exporter snap read the new path with the snap confinement still?

Reply:
Good observation @nobuto-m .

The current implementation is not prepared to have the `golang-openstack-exporter` attached after the upgrade because it will remove but install it again. I will open a new PR to handle this situation.

Thanks

Reply:
Please take a look at the #93 when you have some time to solve this observation of yours.